File: project/CCNet.CSharpDemos.Plugin/Triggers/FileChangedTrigger.cs
using System;
using System.IO;
using Exortech.NetReflector;
using ThoughtWorks.CruiseControl.Core;
using ThoughtWorks.CruiseControl.Core.Config;
using ThoughtWorks.CruiseControl.Core.Triggers;
using ThoughtWorks.CruiseControl.Remote;

namespace CCNet.CSharpDemos.Plugin.Triggers
{
    [ReflectorType("fileChangedTrigger")]
    public class FileChangedTrigger
        : ITrigger, IConfigurationValidation
    {
        private DateTime? lastChanged;

        public FileChangedTrigger()
        {
            this.BuildCondition = BuildCondition.IfModificationExists;
            this.InnerTrigger = new IntervalTrigger { IntervalSeconds = 5 };
        }

        [ReflectorProperty("file")]
        public string MonitorFile { get; set; }

        [ReflectorProperty("buildCondition", Required = false)]
        public BuildCondition BuildCondition { get; set; }

        [ReflectorProperty("trigger", InstanceTypeKey = "type", Required = false)]
        public ITrigger InnerTrigger { get; set; }

        public DateTime NextBuild
        {
            get { return DateTime.MaxValue; }
        }

        public void IntegrationCompleted()
        {
            this.InnerTrigger.IntegrationCompleted();
            this.lastChanged = File.GetLastWriteTime(this.MonitorFile);
        }

        public IntegrationRequest Fire()
        {
            IntegrationRequest request = null;
            if (this.lastChanged.HasValue)
            {
                if (this.InnerTrigger.Fire() != null)
                {
                    var changeTime = File.GetLastWriteTime(this.MonitorFile);
                    if (changeTime > this.lastChanged.Value)
                    {
                        request = new IntegrationRequest(
                            this.BuildCondition,
                            this.GetType().Name,
                            null);
                        this.lastChanged = changeTime;
                    }
                }
            }
            else
            {
                this.lastChanged = File.GetLastWriteTime(this.MonitorFile);
            }

            return request;
        }

        public void Validate(IConfiguration configuration,
            ConfigurationTrace parent,
            IConfigurationErrorProcesser errorProcesser)
        {
            if (string.IsNullOrEmpty(this.MonitorFile))
            {
                errorProcesser.ProcessError("File cannot be empty");
            }
            else if (!File.Exists(this.MonitorFile))
            {
                errorProcesser.ProcessWarning(
                    "File '" + this.MonitorFile + "' does not exist");
            }
        }
    }
}

File: project/CCNet.VBDemos.Plugin/Triggers/FileChangedTrigger.vb
Imports System
Imports System.IO
Imports Exortech.NetReflector
Imports ThoughtWorks.CruiseControl.Core
Imports ThoughtWorks.CruiseControl.Core.Config
Imports ThoughtWorks.CruiseControl.Core.Triggers
Imports ThoughtWorks.CruiseControl.Remote

<ReflectorType("fileChangedTrigger")> Public Class FileChangedTrigger
    Implements ITrigger, IConfigurationValidation

    Private myLastChanged As DateTime?

    Public Sub New()
        BuildCondition = BuildCondition.IfModificationExists
        Dim defaultTrigger = New IntervalTrigger()
        defaultTrigger.IntervalSeconds = 5
        InnerTrigger = defaultTrigger
    End Sub

    Private myMonitorFile As String
    <ReflectorProperty("file")> _
    Public Property MonitorFile() As String
        Get
            Return myMonitorFile
        End Get
        Set(ByVal value As String)
            myMonitorFile = value
        End Set
    End Property

    Private myBuildCondition As BuildCondition
    <ReflectorProperty("buildCondition", Required:=False)> _
    Public Property BuildCondition() As BuildCondition
        Get
            Return myBuildCondition
        End Get
        Set(ByVal value As BuildCondition)
            myBuildCondition = value
        End Set
    End Property

    Private myInnerTrigger As ITrigger
    <ReflectorProperty("trigger", InstanceTypeKey:="type", Required:=False)> _
    Public Property InnerTrigger() As ITrigger
        Get
            Return myInnerTrigger
        End Get
        Set(ByVal value As ITrigger)
            myInnerTrigger = value
        End Set
    End Property

    Public Function Fire() As IntegrationRequest Implements ITrigger.Fire
        Dim request As IntegrationRequest = Nothing
        If myLastChanged.HasValue Then
            If myInnerTrigger.Fire() IsNot Nothing Then
                Dim changeTime = File.GetLastWriteTime(MonitorFile)
                If (changeTime > myLastChanged.Value) Then
                    request = New IntegrationRequest(BuildCondition, _
                                                     Me.GetType().Name, _
                                                     Nothing)
                    myLastChanged = changeTime
                End If
            End If
        Else
            myLastChanged = File.GetLastWriteTime(MonitorFile)
        End If

        Return request
    End Function

    Public Sub IntegrationCompleted() Implements ITrigger.IntegrationCompleted
        InnerTrigger.IntegrationCompleted()
        myLastChanged = File.GetLastWriteTime(MonitorFile)
    End Sub

    Public ReadOnly Property NextBuild() As Date Implements ITrigger.NextBuild
        Get
            Return DateTime.MaxValue
        End Get
    End Property

    Public Sub Validate(ByVal configuration As IConfiguration, _
                        ByVal parent As ConfigurationTrace, _
                        ByVal errorProcesser As IConfigurationErrorProcesser) _
                        Implements IConfigurationValidation.Validate
        If (String.IsNullOrEmpty(myMonitorFile)) Then
            errorProcesser.ProcessError("File cannot be empty")
        ElseIf (Not File.Exists(MonitorFile)) Then
            errorProcesser.ProcessWarning("File '" & MonitorFile & "' does not exist")
        End If
    End Sub
End Class
